src/Data/IntervalRelation.hs

```diff
@@ -0,0 +1,88 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE CPP, DeriveDataTypeable, DeriveGeneric #-}
+{-# LANGUAGE Safe #-}
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.IntervalRelation
+-- Copyright   :  (c) Masahiro Sakai 2016
+-- License     :  BSD-style
+--
+-- Maintainer  :  masahiro.sakai@gmail.com
+-- Stability   :  provisional
+-- Portability :  non-portable (CPP, DeriveDataTypeable, DeriveGeneric)
+--
+-- Interval relations and their algebra.
+--
+-----------------------------------------------------------------------------
+module Data.IntervalRelation
+  ( Relation(..)
+  , relate
+  , invert
+  )
+  where
+
+import Prelude hiding (null)
+import Data.Data
+import Data.Interval
+import GHC.Generics (Generic)
+
+-- | describes how two intervals @x@ and @y@ can be related.
+-- See [Allen's interval algebra](https://en.wikipedia.org/wiki/Allen%27s_interval_algebra)
+data Relation
+  = Equal
+  | Starts
+  | Finishes
+  | During
+  | StartedBy
+  | FinishedBy
+  | Contains
+  | Before
+  | After
+  | JustBefore
+  | JustAfter
+  | Overlaps
+  | OverlappedBy
+  deriving (Eq, Ord, Enum, Bounded, Show, Read, Generic, Data, Typeable)
+
+-- | Computes how two intervals are related according to the @`Relation`@ classification
+relate :: Ord r => Interval r -> Interval r -> Relation
+relate i1 i2 =
+  case (i1 `intersection` i2 == i1, i1 `intersection` i2 == i2) of
```

Comment:
Could you possibly use `Data.Interval.isSubsetOf` and `Data.Interval.isProperSubsetOf` instead of `intersection` here?